src/Illuminate/Database/Eloquent/Builder.php

```diff
@@ -997,6 +997,45 @@ public function create(array $attributes = [])
         });
     }
 
+    /**
+     * Insert new records into the database, casting properties and setting timestamps where applicable.
+     *
+     * @param  iterable  $values
+     * @return bool
+     */
+    public function insertWithCasts(iterable $values): bool
```

Comment:
What about the following setup:

```php
$models = collect();
foreach ($values as $key => $attributes) {
    $model = $this->newModelInstance($attributes);
    if ($model->usesTimestamps()) {
        $model->updateTimestamps();
    }
    $models[] = $model;
}
// Trigger pre-events
$this->toBase()->insert($models->map(fn ($m) => $m->getAttributes());
// Trigger post-events
```

This uses more of the Eloquent methods and is more of a `createMany` function (it isn't that much different from your code actually). Also `getAttributes()` is used internally to create insert statements: https://github.com/laravel/framework/blob/997218b35f5a7f2d5cb7c640e690a5af03499658/src/Illuminate/Database/Eloquent/Model.php#L1297)
(which routes to the public fn `getAttributes()`)

There are some remaining questions with triggering the events though, e.g. what happens when one event returns false? In current situations, the update fails so it might be wise to use the same thing here.

Also the `unguarded()` wrapper can probably be removed since I believe its expected to be guarded by default. You can wrap it yourself in the outer function:
```php
User::unguarded(function() {
    User::createMany($values);
});
```

Just spouting some ideas here

Reply:
@bert-w thanks for the thoughtful comments!

### Timestamps
I wanted to avoid creating a new set of timestamps for something that would essentially be static across all of the models being inserted. Not a make or break, and would be amenable to making that change.  Just seemed like unnecessary overhead.

### Use of unguarded()
This is here so that the call to `Model@fill()` doesn't discard attributes that are being passed into `newModelInstance()`, even if they are guarded. If we were to change it to `createMany` then that'd be great and would happily implement your suggestion. 👍 

The way `Builder@insert()` works is that it let's you pass *anything* to be inserted, guarded or not. I was just keeping with that convention.

#### An aside
There's also an issue I never noticed before, which is if you pass two arrays that have different keys, MySQL throws an exception.
```php
Album::insert([
    ['title' => 'Kind of Blue', 'genre' => 'jazz'],
    ['title' => 'Highway 61 Revisited', 'artist' => 'Bob Dylan', 'genre' => 'rock']
]);
```
```zsh
Illuminate\Database\QueryException  SQLSTATE[21S01]: Insert value list does not match column list: 1136 Column count doesn't match value count at row 2 (Connection: mysql, SQL: insert into `albums` (`genre`, `title`) values (jazz, Kind of Blue), (Bob Dylan, rock, Highway 61 Revisited)).
```

(I feel there should be a PR for adding a DEFAULT value to an insert/update/etc statement, but well beyond the scope of this PR).

##
I guess we'll wait to see what the framework maintainers say. I think this could also be improved by letting the caller pass an array or collection of models that need persisted (in case they need to set the state with their own Model methods or something).

Reply:
there's actually an interesting workaround: you can use last-inserted-id to get the id of the first row of the last batch. Since the batch is one atomic query, the ids are bound to be sequential (assuming mysql+innodb) so the other rows can be simply calculated. But this might introduce inconsistencies across other db drivers since im not sure what happens there (however the integration test should cover it since it runs for all supported db drivers).

Reply:
That's a bit too presumptuous I'm afraid. Within MySQL, you can actually change your [auto-increment increment](https://dev.mysql.com/blog-archive/mysql-group-replication-auto-increment-configuration-handling/) value to something other than 1.

I don't know a lot about PostgreSQL, but I think that engine allows for a [SERIAL column type](https://www.postgresqltutorial.com/postgresql-tutorial/postgresql-serial/) for specifying how you might want to auto-increment.